azd ai agent run on .net projects incorrectly says port 8088

<img width="584" height="317" alt="Image" src="https://github.com/user-attachments/assets/4c89d57a-69e4-42be-baae-6fae953f3901" />

Noticed using the azd ai agent run saying it was in port 8088 but actually the host is running in 5000

Got this error after running.

```
> azd ai agent init
```

Copied a csproj + Program.cs file (without Dockerfile).

```
> azd ai agent run

Detected dotnet project. Start command: dotnet run

After startup, in another terminal, try:
  azd ai agent invoke --local "Hello!"

Starting agent on http://localhost:8088 (Ctrl+C to stop)

D:\repo\what\simple-agent.csproj : warning NU1603: simple-agent depends on Microsoft.Agents.AI.Foundry (>= 0.9.0-hosted.260413.1) but Microsoft.Agents.AI.Foundry 0.9.0-hosted.260413.1 was not found. Microsoft.Agents.AI.Foundry 0.9.0-hosted.local was resolved instead.
D:\repo\what\simple-agent.csproj : warning NU1603: simple-agent depends on Microsoft.Agents.AI.Foundry (>= 0.9.0-hosted.260413.1) but Microsoft.Agents.AI.Foundry 0.9.0-hosted.260413.1 was not found. Microsoft.Agents.AI.Foundry 0.9.0-hosted.local was resolved instead.
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: D:\repo\what
```

```
> azd ai agent invoke --local "Hello!"

Target:       localhost:8088 (local)
Message:      "Hello!"
Session:      d4f71323-b5f0-4cf4-950d-7bfb9a27a70e
Conversation: 44f587bc-7aa5-4143-b4fc-b09c64c82980


ERROR: could not connect to localhost:8088 — is the agent running? Start it with: azd ai agent run
```

--- a/cli/azd/extensions/azure.ai.agents/internal/cmd/run.go
+++ b/cli/azd/extensions/azure.ai.agents/internal/cmd/run.go
@@ -138,13 +138,7 @@ func runRun(ctx context.Context, flags *runFlags) error {
 	cmdParts = resolveVenvCommand(projectDir, cmdParts)
 
 	env := os.Environ()
-	env = append(env, fmt.Sprintf("PORT=%d", flags.port))
-
-	// ASP.NET Core ignores the PORT env var — it uses ASPNETCORE_URLS to configure
-	// Kestrel's listening address. Set it so .NET agents bind to the correct port.
-	if pt.Language == "dotnet" {
-		env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", flags.port))
-	}
+	env = appendPortEnvVars(env, pt, flags.port)
 
 	// Load azd environment variables (e.g., AZURE_AI_PROJECT_ENDPOINT)
 	// so the agent can reach Azure services during local development.
@@ -207,6 +201,17 @@ func runRun(ctx context.Context, flags *runFlags) error {
 		return fmt.Errorf("agent exited: %w", err)
 	}
 	return nil
+}
+
+// appendPortEnvVars appends PORT and, for .NET projects, ASPNETCORE_URLS to the
+// environment slice so the agent listens on the correct port.
+// ASP.NET Core ignores PORT — it uses ASPNETCORE_URLS to configure Kestrel.
+func appendPortEnvVars(env []string, pt ProjectType, port int) []string {
+	env = append(env, fmt.Sprintf("PORT=%d", port))
+	if pt.Language == "dotnet" {
+		env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", port))
+	}
+	return env
 }
 
 // --- Dependency installation ---

--- a/cli/azd/extensions/azure.ai.agents/internal/cmd/run.go
+++ b/cli/azd/extensions/azure.ai.agents/internal/cmd/run.go
@@ -89,6 +89,10 @@ func runRun(ctx context.Context, flags *runFlags) error {
 	}
 	projectDir := runCtx.ProjectDir
 
+	// Detect project type early — used for both start-command resolution and
+	// environment setup (e.g., setting ASPNETCORE_URLS for .NET).
+	pt := detectProjectType(projectDir)
+
 	// Resolve start command: --start-command flag > azure.yaml startupCommand > detect
 	startCmd := flags.startCommand
 	if startCmd == "" {
@@ -96,7 +100,6 @@ func runRun(ctx context.Context, flags *runFlags) error {
 	}
 
 	if startCmd == "" {
-		pt := detectProjectType(projectDir)
 		if pt.StartCmd != "" {
 			startCmd = pt.StartCmd
 			fmt.Printf("Detected %s project. Start command: %s\n", pt.Language, startCmd)
@@ -136,6 +139,12 @@ func runRun(ctx context.Context, flags *runFlags) error {
 
 	env := os.Environ()
 	env = append(env, fmt.Sprintf("PORT=%d", flags.port))
+
+	// ASP.NET Core ignores the PORT env var — it uses ASPNETCORE_URLS to configure
+	// Kestrel's listening address. Set it so .NET agents bind to the correct port.
+	if pt.Language == "dotnet" {
+		env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", flags.port))
+	}
 
 	// Load azd environment variables (e.g., AZURE_AI_PROJECT_ENDPOINT)
 	// so the agent can reach Azure services during local development.

--- a/cli/azd/extensions/azure.ai.agents/internal/cmd/run_test.go
+++ b/cli/azd/extensions/azure.ai.agents/internal/cmd/run_test.go
@@ -4,6 +4,7 @@
 package cmd
 
 import (
+	"fmt"
 	"os"
 	"path/filepath"
 	"runtime"
@@ -333,6 +334,138 @@ func TestAppendFoundryEnvVars(t *testing.T) {
 		}
 		if !slices.Contains(env, "FOUNDRY_AGENT_VERSION=v2") {
 			t.Errorf("expected FOUNDRY_AGENT_VERSION to be translated, got %v", env)
+		}
+	})
+}
+
+func TestDotnetProjectSetsAspnetcoreUrls(t *testing.T) {
+	t.Parallel()
+
+	t.Run("dotnet project includes ASPNETCORE_URLS", func(t *testing.T) {
+		t.Parallel()
+		dir := t.TempDir()
+		// Create a .csproj file so detectProjectType returns "dotnet"
+		if err := os.WriteFile(
+			filepath.Join(dir, "MyAgent.csproj"),
+			[]byte("<Project/>"),
+			0600,
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		pt := detectProjectType(dir)
+		if pt.Language != "dotnet" {
+			t.Fatalf("expected dotnet, got %s", pt.Language)
+		}
+
+		port := 8088
+		var env []string
+		env = append(env, fmt.Sprintf("PORT=%d", port))
+		if pt.Language == "dotnet" {
+			env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", port))
+		}
+
+		if !slices.Contains(env, "PORT=8088") {
+			t.Errorf("expected PORT=8088 in env, got %v", env)
+		}
+		if !slices.Contains(env, "ASPNETCORE_URLS=http://localhost:8088") {
+			t.Errorf("expected ASPNETCORE_URLS=http://localhost:8088 in env, got %v", env)
+		}
+	})
+
+	t.Run("python project does not include ASPNETCORE_URLS", func(t *testing.T) {
+		t.Parallel()
+		dir := t.TempDir()
+		if err := os.WriteFile(
+			filepath.Join(dir, "requirements.txt"),
+			[]byte("flask"),
+			0600,
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		pt := detectProjectType(dir)
+		if pt.Language != "python" {
+			t.Fatalf("expected python, got %s", pt.Language)
+		}
+
+		port := 8088
+		var env []string
+		env = append(env, fmt.Sprintf("PORT=%d", port))
+		if pt.Language == "dotnet" {
+			env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", port))
+		}
+
+		if !slices.Contains(env, "PORT=8088") {
+			t.Errorf("expected PORT=8088 in env, got %v", env)
+		}
+		for _, entry := range env {
+			if strings.HasPrefix(entry, "ASPNETCORE_URLS=") {
+				t.Errorf("ASPNETCORE_URLS should not be set for python projects, got %v", env)
+			}
+		}
+	})
+
+	t.Run("node project does not include ASPNETCORE_URLS", func(t *testing.T) {
+		t.Parallel()
+		dir := t.TempDir()
+		if err := os.WriteFile(
+			filepath.Join(dir, "package.json"),
+			[]byte("{}"),
+			0600,
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		pt := detectProjectType(dir)
+		if pt.Language != "node" {
+			t.Fatalf("expected node, got %s", pt.Language)
+		}
+
+		port := 9090
+		var env []string
+		env = append(env, fmt.Sprintf("PORT=%d", port))
+		if pt.Language == "dotnet" {
+			env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", port))
+		}
+
+		if !slices.Contains(env, "PORT=9090") {
+			t.Errorf("expected PORT=9090 in env, got %v", env)
+		}
+		for _, entry := range env {
+			if strings.HasPrefix(entry, "ASPNETCORE_URLS=") {
+				t.Errorf("ASPNETCORE_URLS should not be set for node projects, got %v", env)
+			}
+		}
+	})
+
+	t.Run("dotnet project respects custom port", func(t *testing.T) {
+		t.Parallel()
+		dir := t.TempDir()
+		if err := os.WriteFile(
+			filepath.Join(dir, "MyAgent.csproj"),
+			[]byte("<Project/>"),
+			0600,
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		pt := detectProjectType(dir)
+		port := 3000
+		var env []string
+		env = append(env, fmt.Sprintf("PORT=%d", port))
+		if pt.Language == "dotnet" {
+			env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", port))
+		}
+
+		if !slices.Contains(env, "PORT=3000") {
+			t.Errorf("expected PORT=3000 in env, got %v", env)
+		}
+		if !slices.Contains(env, "ASPNETCORE_URLS=http://localhost:3000") {
+			t.Errorf(
+				"expected ASPNETCORE_URLS=http://localhost:3000 in env, got %v",
+				env,
+			)
 		}
 	})
 }

--- a/cli/azd/extensions/azure.ai.agents/internal/cmd/run_test.go
+++ b/cli/azd/extensions/azure.ai.agents/internal/cmd/run_test.go
@@ -4,7 +4,6 @@
 package cmd
 
 import (
-	"fmt"
 	"os"
 	"path/filepath"
 	"runtime"
@@ -338,32 +337,13 @@ func TestAppendFoundryEnvVars(t *testing.T) {
 	})
 }
 
-func TestDotnetProjectSetsAspnetcoreUrls(t *testing.T) {
+func TestAppendPortEnvVars(t *testing.T) {
 	t.Parallel()
 
 	t.Run("dotnet project includes ASPNETCORE_URLS", func(t *testing.T) {
 		t.Parallel()
-		dir := t.TempDir()
-		// Create a .csproj file so detectProjectType returns "dotnet"
-		if err := os.WriteFile(
-			filepath.Join(dir, "MyAgent.csproj"),
-			[]byte("<Project/>"),
-			0600,
-		); err != nil {
-			t.Fatal(err)
-		}
-
-		pt := detectProjectType(dir)
-		if pt.Language != "dotnet" {
-			t.Fatalf("expected dotnet, got %s", pt.Language)
-		}
-
-		port := 8088
-		var env []string
-		env = append(env, fmt.Sprintf("PORT=%d", port))
-		if pt.Language == "dotnet" {
-			env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", port))
-		}
+		pt := ProjectType{Language: "dotnet", StartCmd: "dotnet run"}
+		env := appendPortEnvVars(nil, pt, 8088)
 
 		if !slices.Contains(env, "PORT=8088") {
 			t.Errorf("expected PORT=8088 in env, got %v", env)
@@ -375,97 +355,59 @@ func TestDotnetProjectSetsAspnetcoreUrls(t *testing.T) {
 
 	t.Run("python project does not include ASPNETCORE_URLS", func(t *testing.T) {
 		t.Parallel()
-		dir := t.TempDir()
-		if err := os.WriteFile(
-			filepath.Join(dir, "requirements.txt"),
-			[]byte("flask"),
-			0600,
-		); err != nil {
-			t.Fatal(err)
-		}
-
-		pt := detectProjectType(dir)
-		if pt.Language != "python" {
-			t.Fatalf("expected python, got %s", pt.Language)
-		}
-
-		port := 8088
-		var env []string
-		env = append(env, fmt.Sprintf("PORT=%d", port))
-		if pt.Language == "dotnet" {
-			env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", port))
-		}
+		pt := ProjectType{Language: "python", StartCmd: "python main.py"}
+		env := appendPortEnvVars(nil, pt, 8088)
 
 		if !slices.Contains(env, "PORT=8088") {
 			t.Errorf("expected PORT=8088 in env, got %v", env)
 		}
 		for _, entry := range env {
 			if strings.HasPrefix(entry, "ASPNETCORE_URLS=") {
-				t.Errorf("ASPNETCORE_URLS should not be set for python projects, got %v", env)
+				t.Errorf("ASPNETCORE_URLS should not be set for python, got %v", env)
 			}
 		}
 	})
 
 	t.Run("node project does not include ASPNETCORE_URLS", func(t *testing.T) {
 		t.Parallel()
-		dir := t.TempDir()
-		if err := os.WriteFile(
-			filepath.Join(dir, "package.json"),
-			[]byte("{}"),
-			0600,
-		); err != nil {
-			t.Fatal(err)
-		}
-
-		pt := detectProjectType(dir)
-		if pt.Language != "node" {
-			t.Fatalf("expected node, got %s", pt.Language)
-		}
-
-		port := 9090
-		var env []string
-		env = append(env, fmt.Sprintf("PORT=%d", port))
-		if pt.Language == "dotnet" {
-			env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", port))
-		}
+		pt := ProjectType{Language: "node", StartCmd: "npm start"}
+		env := appendPortEnvVars(nil, pt, 9090)
 
 		if !slices.Contains(env, "PORT=9090") {
 			t.Errorf("expected PORT=9090 in env, got %v", env)
 		}
 		for _, entry := range env {
 			if strings.HasPrefix(entry, "ASPNETCORE_URLS=") {
-				t.Errorf("ASPNETCORE_URLS should not be set for node projects, got %v", env)
+				t.Errorf("ASPNETCORE_URLS should not be set for node, got %v", env)
 			}
 		}
 	})
 
 	t.Run("dotnet project respects custom port", func(t *testing.T) {
 		t.Parallel()
-		dir := t.TempDir()
-		if err := os.WriteFile(
-			filepath.Join(dir, "MyAgent.csproj"),
-			[]byte("<Project/>"),
-			0600,
-		); err != nil {
-			t.Fatal(err)
-		}
-
-		pt := detectProjectType(dir)
-		port := 3000
-		var env []string
-		env = append(env, fmt.Sprintf("PORT=%d", port))
-		if pt.Language == "dotnet" {
-			env = append(env, fmt.Sprintf("ASPNETCORE_URLS=http://localhost:%d", port))
-		}
+		pt := ProjectType{Language: "dotnet", StartCmd: "dotnet run"}
+		env := appendPortEnvVars(nil, pt, 3000)
 
 		if !slices.Contains(env, "PORT=3000") {
 			t.Errorf("expected PORT=3000 in env, got %v", env)
 		}
-		if !slices.Contains(env, "ASPNETCORE_URLS=http://localhost:3000") {
-			t.Errorf(
-				"expected ASPNETCORE_URLS=http://localhost:3000 in env, got %v",
-				env,
-			)
+		expected := "ASPNETCORE_URLS=http://localhost:3000"
+		if !slices.Contains(env, expected) {
+			t.Errorf("expected %q in env, got %v", expected, env)
+		}
+	})
+
+	t.Run("preserves existing env entries", func(t *testing.T) {
+		t.Parallel()
+		pt := ProjectType{Language: "dotnet", StartCmd: "dotnet run"}
+		existing := []string{"HOME=/home/user", "PATH=/usr/bin"}
+		env := appendPortEnvVars(existing, pt, 8088)
+
+		if len(env) != 4 {
+			t.Errorf("expected 4 entries (2 existing + PORT + ASPNETCORE_URLS), got %d: %v", len(env), env)
+		}
+		if env[0] != "HOME=/home/user" || env[1] != "PATH=/usr/bin" {
+			t.Errorf("existing entries not preserved, got %v", env)
 		}
 	})
 }

